pkg/pgclient/client.go

```diff
@@ -33,84 +34,108 @@ type LockFunc = func(ctx context.Context, conn *pgx.Conn) error
 
 // Client sends Prometheus samples to TimescaleDB
 type Client struct {
-	Connection        pgxconn.PgxConn
-	QuerierConnection pgxconn.PgxConn
-	ingestor          *ingestor.DBIngestor
-	querier           querier.Querier
-	promqlEngine      *promql.Engine
-	healthCheck       health.HealthCheckerFn
-	queryable         promql.Queryable
-	ConnectionStr     string
-	metricCache       cache.MetricCache
-	labelsCache       cache.LabelsCache
-	seriesCache       cache.SeriesCache
-	closePool         bool
-	sigClose          chan struct{}
-	haService         *ha.Service
+	readerConn   pgxconn.PgxConn
+	ingestor     *ingestor.DBIngestor
+	querier      querier.Querier
+	promqlEngine *promql.Engine
+	healthCheck  health.HealthCheckerFn
+	queryable    promql.Queryable
+	metricCache  cache.MetricCache
+	labelsCache  cache.LabelsCache
+	seriesCache  cache.SeriesCache
+	connPools    []pgxconn.PgxConn
+	sigClose     chan struct{}
+	haService    *ha.Service
 }
 
 // NewClient creates a new PostgreSQL client
-func NewClient(cfg *Config, mt tenancy.Authorizer, schemaLocker LockFunc, readOnly bool) (*Client, error) {
-	pgConfig, numCopiers, err := getPgConfig(cfg)
+func NewClient(r prometheus.Registerer, cfg *Config, mt tenancy.Authorizer, schemaLocker LockFunc, readOnly bool) (*Client, error) {
```

Comment:
Would it make sense to refactor this into 2 different functions? Something like

- NewClient
- NewReadOnlyClient

And moving the code that's inside the `!readOnly` into a `getWriterPool` function and the other part into a `getReaderPool` function then the Client constructors will just call this functions and pass the values to `NewClientWithPool`

Reply:
This is a good idea. But, given the current PR, it will be out of scope plus a lot of changes plus the existing stuff in this PR. Would you like to work on this refactoring once we merge this PR?

Reply:
Let's create an issue for the refactoring. @alejandrodnm would you mind making it since it is your proposal? :smile:

Reply:
I knew it was a trap 😄 Yeah, I'll do it